Refactors activities using google sheets and google drive

shobhitagarwal1612 opened the pull request:
Adds GoogleAccountManager class to handle google account selection and permission
Adds DriveHelper to interact with Google Drive
Adds SheetsHelper to interact with Google Sheets
Uses above classes in GoogleDriveActivity and GoogleSheetsUploaderActivity

Closes #1568

#### What has been done to verify that this works as intended?
 - Downloading a form from GoogleDriveActivity
 - Account picker dialog appears if no account is selected in Settings
 - Permission dialog appears if the application does not have authorization permissions
 - Successfully uploading forms to google sheets (multiple at a time)
 - Successfully uploading media to google drive which is in the form instance

#### Why is this the best possible solution? Were any other approaches considered?
Improvements are always welcome

#### Are there any risks to merging this code? If so, what are they?
Yes, all of the functionalities need to be retested

#### Do we need any specific form for testing your changes? If so, please attach one.
I used the all widgets form for testing

Commit:
Adds GoogleAccountManager class to handle google account selection and permission

Adds DriveHelper to interact with Google Drive

Adds SheetsHelper to interact with Google Sheets

Uses above classes in GoogleDriveActivity and GoogleSheetsUploaderActivity

Commit:
Fixes PMD issues

Commit:
Fixes checkstyle issues

Commit:
Merge branch 'master' into refactor-sheets-drive

# Conflicts:
#	collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java

Commit:
Refactor the latest merged code for downloading media files

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+    public boolean downloadFile(DriveListItem fileItem, HashMap<String, Object> results) {
+        FileOutputStream fileOutputStream = null;
+        try {
+            fileOutputStream = new FileOutputStream(new File(Collect.FORMS_PATH + File.separator + fileItem.getName()));
```
This should be provided by an injected class so we can mock out the FileOutputStream as needed.

heyjamesknight replied:
Or at the very least, by a separate method so we can do a partial mock/spy (i.e. keep the default implementation for everything but these provider methods, which we override through mocking)

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -123,18 +114,20 @@ protected void onCreate(Bundle savedInstanceState) {
-        credential = GoogleAccountCredential.usingOAuth2(
-                getApplicationContext(), Collections.singleton(DriveScopes.DRIVE))
-                .setBackOff(new ExponentialBackOff());
+        accountsManager = new GoogleAccountsManager(this, new GoogleAccountsManager.PermissionsListener() {
```
Could we make the Activity implement PermissionsListener and kill the anonymous class?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -163,53 +156,22 @@ private void runTask() {
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        accountsManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
```
Good idea moving this out.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+    private Drive drive;
+
+    DriveHelper(GoogleAccountCredential credential, HttpTransport transport, JsonFactory jsonFactory) {
+        drive = new Drive.Builder(transport, jsonFactory, credential)
```
Should provide a second constructor that takes the single Drive object so a mock can be passed in.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+
+public class DriveHelper {
+
+    private Drive drive;
```
Can be final.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+
+    private ByteArrayOutputStream downloadFile(String fileId) throws IOException {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        drive.files().get(fileId).executeMediaAndDownloadTo(outputStream);
```
Is there any reason we don't write this directly to the FileOutputStream, rather than first to a ByteArrayOutputStream?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+        return getIDOfFolderWithName(GOOGLE_DRIVE_SUBFOLDER, rootFolderId);
+    }
+
+    public String getIDOfFolderWithName(String name, String inFolder) throws IOException, MultipleFoldersFoundException {
```
What is `inFolder`? Can you add a Javadoc comment explaining what the parameters are?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+        com.google.api.services.drive.model.File folder = null;
+
+        // check if the folder exists
+        ArrayList<com.google.api.services.drive.model.File> files = getFilesFromDrive(name, inFolder);
```
Prefer interfaces of implementations for declarations when possible (e.g. List not ArrayList)

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+            if (folder == null) {
+                folder = file;
+            } else {
+                throw new MultipleFoldersFoundException("Multiple \"" + name + "\" folders found");
```
This is a confusing way to do this. Can we just check if `files.size()` is greater than 1 and throw the exception there?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+
+    }
+
+    public String uploadFileToDrive(String mediaName, String destinationFolderID,
```
Can we add a javadoc comment describing what the parameters are?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,181 @@
+    private static final int RESOLVE_CONNECTION_REQUEST_CODE = 5555;
+
+    private Context context;
+    private Activity activity;
```
This should be marked `@Nullable`, and we should internally watch for those null cases.

heyjamesknight replied:
These are nullable, but we access them a few times without checking that. We should add a `@Nullable` here and check that to avoid NPEs.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,181 @@
+                listener.accountSelected();
+            } else {
+                // Start a dialog from which the user can choose an account
+                activity.startActivityForResult(credential.newChooseAccountIntent(),
```
Currently someone could create a `GoogleAccountsManager` with the `Context` constructor, and call `chooseAccount`, which would result in an NPE.

heyjamesknight replied:
Alternatively you could just hang onto the Context from the Constructor and manually pass in the Activity to this function.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,181 @@
+            try {
+                connectionResult.startResolutionForResult(activity, RESOLVE_CONNECTION_REQUEST_CODE);
+            } catch (IntentSender.SendIntentException e) {
+                // Unable to resolve, message user appropriately
```
Is this unfinished?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+        return null;
+    }
+
+    public boolean downloadFile(DriveListItem fileItem, HashMap<String, Object> results) {
```
What is results? What does a true/false return value here mean?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+        // check if the folder exists
+        ArrayList<com.google.api.services.drive.model.File> files = getFilesFromDrive(name, inFolder);
+
+        for (com.google.api.services.drive.model.File file : files) {
```
This is a confusing way to do this: can we just check if `files.size() > 1` instead?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+
+    }
+
+    public String uploadFileToDrive(String mediaName, String destinationFolderID,
```
What are the parameters here? What do they mean?

heyjamesknight replied:
Also, what is the String that is being returned?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+        return folder;
+    }
+
+    private ArrayList<com.google.api.services.drive.model.File> getFilesFromDrive(
```
Same here, what are these parameters?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,216 @@
+                .getId();
+    }
+
+    private com.google.api.services.drive.model.File createFolderInDrive(String folderName,
```
Same here, what are these parameters?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+
+    public DriveHelper getDriveHelper() {
```
These are great because you can use them to provide mocked implementations.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,181 @@
+                listener.accountSelected();
+            } else {
+                // Start a dialog from which the user can choose an account
+                activity.startActivityForResult(credential.newChooseAccountIntent(),
```
NPE risk.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/SheetsHelper.java:
```diff
@@ -0,0 +1,146 @@
+     * to modify the given spreadsheetId. If yes, then returns complete spreadsheet
+     * otherwise throws exception
+     */
+    public Spreadsheet checkPermissions(String spreadsheetId) throws IOException {
```
`updateSpreadsheetPermissions` might be a better name here, as its a bit confusing as to why we're returning a Spreadsheet from a method titled `check`.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/SheetsHelper.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Inserts a new row in the given sheet of the spreadsheet
+     */
+    public void insertRow(String spreadsheetId, String sheetName, ValueRange row) throws IOException {
```
@shobhitagarwal1612 this is a good example of the type of method I mention in my overall comment.

heyjamesknight reviewed (changes requested):
Great job overall. The biggest thing we can do to make this more testable is to move all of the Drive interaction into single-use methods that we can mock out at higher levels of abstraction. That way we can test any interaction between the app and Drive separate from all of the additional logic we have surrounding what we do with the results of that interaction.

I've called out a specific method that's a good example of what I'm suggesting below. I don't think you need to create a new layer of abstraction for this—just moving the calls into their own methods within DriveHelper/SheetsHelper should be enough. 

Once they're moved out, you can create a DriveHelper, then call Mockito.spy() to wrap it in a spy. This will then let you override specific methods with mocked functionality while keeping the rest of the implementation the same. Let me know if you need any help getting that set up.

Commit:
Merge branch 'master' into refactor-sheets-drive

# Conflicts:
#	collect_app/src/main/java/org/odk/collect/android/activities/GoogleDriveActivity.java
#	collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java

Commit:
Overrides theme of account picker dialog

Commit:
Removes duplicate code in ServerPreferencesFragment

Commit:
Adds javadoc and other minor code refactors

Commit:
Removes parameter from chooseAccount method and converts into a field

shobhitagarwal1612 commented:
@jknightco Please review again. I have fixed all of the issues pointed out by you

Commit:
Removes listener from constructor

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,248 @@
+     *                               whether to create a new folder with the given name or not
+     */
+    @Nullable
+    public String getIDOfFolderWithName(@NonNull String name, @Nullable String inFolder, boolean shouldCreateIfNotFound)
```
You may want to consider having this class *only* make calls to Drive, e.g. these ones:

```
drive.files().get("root") .setFields("id") .execute().getId();
```

We can then have another class that contains business logic like this method—that way we can mock the class out that makes the Drive calls and test this logic here separately.

Not 100% necessary, but something to think about at least.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,248 @@
+     * @param parentId   The id of the folder in which we want to create the new folder
+     * @return the folder object if created successfully
+     */
+    private com.google.api.services.drive.model.File createFolderInDrive(@NonNull String folderName,
```
This method does three separate things: 1. Create and prepare a new File Metadata, 2. Make a Drive call to create a Folder, and 3. Make a Drive call to set permissions. These should probably be three different methods.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,248 @@
+
+        String requestString;
+        if (folderName != null && parentId == null) {
+            requestString = "name = '" + folderName + "' and "
```
Use `String.format` here instead of '+'

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,248 @@
+     *                   If folderName is null then all files are fetched from the drive
+     * @param parentId   (optional) The id of the parent folder containing the given folder
+     */
+    public List<com.google.api.services.drive.model.File> getFilesFromDrive(
```
Same as above, we should probably break out the request string creation and the actual Drive request. That way we can test them separately.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,227 @@
+        this.context = context;
+
+        credential = GoogleAccountCredential
+                .usingOAuth2(context, Collections.singletonList(DriveScopes.DRIVE))
```
Good use of .singletonList!

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    public interface GoogleAccountSelectionListener {
+        void googleAccountSelected();
```
This should be `onGoogleAccountSelected()`

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:
```diff
@@ -354,21 +346,15 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
 
+    @Override
+    public void googleAccountSelected() {
+        String accountName = accountsManager.getCredentials().getSelectedAccountName();
```
Instead of reaching into the accountsManager, we could just pass the String via this call: `googleAccountSelected(String accountName)` as we have access to it at the call site.

heyjamesknight commented:
@shobhitagarwal1612  For testing, here's what I would recommend:
1. Move all of the Drive calls into their own class. This way we can test them one at a time and not involve any other logic in those tests. We just want to verify that they work as they should.
2. Write an Instrumentation Test (in `androidTest`) that loads any Activity, passes it as the Context, and tests the results of the Drive calls. These tests will take a long time and require an actual device/emulator, so putting they'll have to be Instrumentation tests.
3. With the Drive calls 100% isolated, you can then go about writing basic Unit Tests for the remaining business logic.

Let me know if you have any questions!

Commit:
Breaks down methods into smaller ones in DriveHelper
Code refactor

Commit:
Moves all of the Drive calls into DriveService

Commit:
Checkstyle fixes

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/DriveService.java:
```diff
@@ -0,0 +1,83 @@
+ * @author Shobhit Agarwal
+ */
+
+public class DriveService {
```
Perfect! Now we can test this and then mock it out everywhere else so we don't actually need to reach out to Drive.

shobhitagarwal1612 replied:
Should this `DriveService` be made an inner class of `DriveHelper`?

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/SheetsHelper.java:
```diff
@@ -0,0 +1,163 @@
+        requests.add(new Request().setUpdateSheetProperties(updateSheetPropertyRequest));
+
+        // send the API request
+        sheets.spreadsheets()
```
Maybe add a SheetsService too?

shobhitagarwal1612 replied:
:+1:

shobhitagarwal1612 commented:
@jknightco Thank you so much. No questions for now but will keep you notified if I get into some issue while writing tests.

heyjamesknight commented:
Sounds good, I'll be on the lookout :)

Commit:
Adds dependencies for power mock
It is added to mock final methods which is not handled by Mockito

Commit:
Adds tests for GoogleAccountsManager

lognaturel commented:
Odd that the merge conflict is only showing up now! With the new version of `gradle` we need to use `testImplementation` instead of `testCompile.`

Commit:
Adds some more tests

Commit:
Merge branch 'master' into refactor-sheets-drive

# Conflicts:
#	collect_app/build.gradle

Commit:
Remove duplicate dependencies added during merge conflict resolution

Commit:
Reverts the version of mockito-core to 2.8.47 to fix tests

Commit:
Adds tests for DriveHelper

Commit:
PMD fixes

Commit:
Fixes checkstyle issues

Commit:
Adds more tests

shobhitagarwal1612 commented:
@jknightco @lognaturel Please review

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java:
```diff
@@ -229,73 +187,21 @@ protected void onActivityResult(
-    @Override
-    public void onPermissionsDenied(int requestCode, List<String> list) {
-        // Do nothing.
+        if (resultCode == RESULT_CANCELED) {
```
Not a big deal, but consider moving this before the switch(requestCode) block, and returning.

shobhitagarwal1612 replied:
Good idea

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,249 @@
+    /**
+     * This constructor should be used only for testing purposes
+     */
+    public GoogleAccountsManager(@NonNull GoogleAccountCredential credential,
```
Assuming your tests are in the same package as this, you can make this package private. If they're not, consider moving them. 
[Edit: 'package private' is the default visibility level, i.e. what happens if you don't put public/protected/private].

shobhitagarwal1612 replied:
Ok

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,249 @@
+    private SheetsHelper sheetsHelper;
+
+    private HttpTransport transport;
+    private JsonFactory jsonFactory;
```
There are technically code paths where these are not initialized (the testing-only constructor). Technically as long as that constructor is only used for testing you're fine, but I'd still probably like to see @Nullables on these.

shobhitagarwal1612 replied:
:+1:

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,249 @@
+
+    public void showAccountPickerDialog() {
+        intentChooseAccount.putExtra("overrideTheme", 1);
+        intentChooseAccount.putExtra("overrideCustomTheme", 0);
```
These should be constants that explain their values.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+
+    public boolean hasPermissions() {
```
Its not clear from this method's name that it has a side effect (that is checks for permissions). I would swap 'hasPermissions' which implies an accessor and 'checkAccountPermissions' which implies an action, or consider other names.

heyjamesknight commented on collect_app/src/test/java/org/odk/collect/android/google/DriveHelperTest.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Test
+    public void getRootIdTest() throws IOException {
```
Typically test names describe what they're testing, i.e. "getRootIdShouldReturnTheProperRootFolderId".

heyjamesknight commented on collect_app/src/test/java/org/odk/collect/android/google/GoogleAccountsManagerTest.java:
```diff
@@ -0,0 +1,192 @@
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                currentAccount = invocation.getArgument(0);
+                return null;
```
Very nice! These things can get tricky, well done!

heyjamesknight reviewed (changes requested):
Overall the changes and the tests look really great. Well done Shobhit!

I've added some minor change requests.

Commit:
Moves the activity finishing canceled action above switch case

Commit:
Marks fields as nullable
Reduces scope of methods to package-private

Commit:
Removed method hasPermissions

Commit:
Replaces constants with variables

Commit:
Adds sheets service

Commit:
Adds tests for sheets helper

Commit:
Removes unused field

Commit:
Replaces name of tests with more detailed name

shobhitagarwal1612 commented:
@jknightco @lognaturel PR is ready for review

heyjamesknight commented on collect_app/src/test/java/org/odk/collect/android/google/GoogleAccountsManagerTest.java:
```diff
@@ -150,7 +150,7 @@ public void displayAccountPickerDialogInAutoChooseWhenNoAccountIsNotAvailableTes
 
     @Test
-    public void permissionsNotAvailableTest() {
+    public void shouldRequestForPermissionIfPermissionNotGivenTest() {
```
Perfect! Technically we don't need 'Test' at the end anymore (since JUnit4) but there's no need to change that now.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/google/SheetsHelper.java:
```diff
@@ -55,6 +55,13 @@
 
     public void resizeSpreadSheet(String spreadsheetId, int sheetId, int columnSize) throws IOException {
+        if (sheetId < 0) {
+            throw new IllegalArgumentException("Sheet Id should be greater than or equal to 0");
```
Any time we throw Exceptions we should detail what would cause them in a Javadoc comment. Under what conditions is one expected, etc.

heyjamesknight commented on collect_app/src/test/java/org/odk/collect/android/google/SheetsHelperTest.java:
```diff
@@ -0,0 +1,92 @@
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({Spreadsheet.class, SpreadsheetProperties.class})
+public class SheetsHelperTest {
```
This looks great!

heyjamesknight reviewed (approved):
Approved, pending comment addition on the SheetsHelper methods that throw Exceptions.

Everything else looks great. Really well done @shobhitagarwal1612 .

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/google/DriveHelper.java:
```diff
@@ -0,0 +1,274 @@
+        return query.toString();
+    }
+
+    public String getMediaDirName(@NonNull String fileName) {
```
At some point this should be consolidated with `FileUtils.constructMediaPath`

shobhitagarwal1612 replied:
Should I make this change in this PR?

lognaturel replied:
If you see a good quick way to do it. I think I said "at some point" because I realized they have different signatures or something like that.

lognaturel commented:
@jknightco thank you for the thoughtful review and @shobhitagarwal1612, as always, thanks for taking on big, important things. 

I noticed a regression around form media -- it doesn't look like it gets downloaded anymore. I did a quick round of debugging and I see that the media folder ID is fetched correctly but that the folder's contents aren't being pulled. I see the call to `driveHelper.getFilesFromDrive` resulting in an empty list.

Let's do a thorough check of the Drive integration behavior to make sure there are no other such regressions. @mmarciniak90 or @kkrawczyk123, could one of you please help with that?

One small stylistic thing -- I don't love the `google` package name because it feels too general. Naming is hard! There seems to be a few different package naming strategies going on -- by class kind (e.g. activities, fragments), by functional area (e.g. location, external), by layer (e.g. database; or maybe this is also by functional area?). If we follow the class kind approach this could go in `utilities/gdrive`. By functional area it would maybe be called `upload` or `gdrive`? `google` seems its by library vendor which seems unlike any other existing convention. What do others think?

So to recap:
- [x] Add exception comment
- [x] Fix form media regression
- [ ] Run through Drive testing checklist
- [x] Reconsider `google` package name

Getting close!

shobhitagarwal1612 commented:
@lognaturel Thank you for your careful and thoughtful reviews. It seems I forgot to test that case (downloading media along with the form), so thank you again. 

>  Reconsider `google` package name

I agree that we should rename it something more appropriate. I think naming it `utilities/gdrive` makes more sense as the helpers are for uploading to and downloading from google drive.
In addition to that, I also think that moving the `DriveService` and `SheetsService` classes within each `DriveHelper` and `SheetsHelper` as inner classes would be more appropriate as they are not used by any other class. What does everything think?

Commit:
Fixes downloading media files from the media directory

Issue : Folder mime type was getting appended even when the folder name was null.
So fixed it by adding a condition while determining mime type

Commit:
updates the test condition for getFilesFromDrive()
if we don't provide any parent folder id or the folder name, the drive service shouldn't fetch any files from the google drive

Commit:
Adds comment for the thrown exception

lognaturel commented:
I'm happy with `utilities/gdrive` for now! Using inner classes also sounds good. 👍

shobhitagarwal1612 commented:
:+1:

Commit:
Moves the classes under utilities/gdrive

Commit:
DriveService and SheetsService classes made inner classes of helpers

Commit:
Moves test classes from /google to /utilities/gdrive

Commit:
Replaces getMediaName with the existing FileUtils method constructMediaPath
Uses FileUtils.createMediaPath in GoogleDriveActivity
Moves the test from DriveHelperTest to FileUtilsTest

shobhitagarwal1612 commented:
@lognaturel Done

mmarciniak90 commented:
**Tested with success!**
Verified on Android: 4.1, 4.4, 5.1, 6.0, 7.0, 8.0

Verified cases: 

- Downloading a form from Google Drive
- Downloading a form with media from Google Drive
- Account picker dialog appears on Get Blank Form page if no account is selected in Settings 
- Account picker dialog appears on Send Finalized Form page if no account is selected in Settings 
- Successfully uploading multiple at a time forms to google sheets 
- Successfully uploading forms to submission URL defined in form
- Successfully uploading media to submission URL defined in form
- Successfully uploading forms to Fallback submission URL
- Successfully uploading media to Fallback submission URL
- Successfully uploading forms with Cascading Select

Additionally those changes resolve bug which I found in 1.12.2:
* Google account was not visible in `General Settings` -> `Server` when user added it on `Send Finalized Form` or `Get  Blank Form` view

I noticed one bug during testing: https://github.com/opendatakit/collect/issues/1777

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"
